c_emulator/riscv_sim.cpp

```diff
@@ -915,69 +760,15 @@ int main(int argc, char **argv)
     exit(1);
   }
 
-#ifdef RVFI_DII
-  uint64_t entry;
-  if (rvfi_dii) {
-    entry = 0x80000000;
-    int listen_sock = socket(AF_INET, SOCK_STREAM, 0);
-    if (listen_sock == -1) {
-      fprintf(stderr, "Unable to create socket: %s\n", strerror(errno));
-      return 1;
-    }
-    int reuseaddr = 1;
-    if (setsockopt(listen_sock, SOL_SOCKET, SO_REUSEADDR, &reuseaddr,
-                   sizeof(reuseaddr))
-        == -1) {
-      fprintf(stderr, "Unable to set reuseaddr on socket: %s\n",
-              strerror(errno));
-      return 1;
-    }
-    struct sockaddr_in addr;
-    addr.sin_family = AF_INET;
-    addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
-    addr.sin_port = htons(rvfi_dii_port);
-    if (bind(listen_sock, (struct sockaddr *)&addr, sizeof(addr)) == -1) {
-      fprintf(stderr, "Unable to set bind socket: %s\n", strerror(errno));
-      return 1;
-    }
-    if (listen(listen_sock, 1) == -1) {
-      fprintf(stderr, "Unable to listen on socket: %s\n", strerror(errno));
-      return 1;
-    }
-    socklen_t addrlen = sizeof(addr);
-    if (getsockname(listen_sock, (struct sockaddr *)&addr, &addrlen) == -1) {
-      fprintf(stderr, "Unable to getsockname() on socket: %s\n",
-              strerror(errno));
-      return 1;
-    }
-    printf("Waiting for connection on port %d.\n", ntohs(addr.sin_port));
-    rvfi_dii_sock = accept(listen_sock, NULL, NULL);
-    if (rvfi_dii_sock == -1) {
-      fprintf(stderr, "Unable to accept connection on socket: %s\n",
-              strerror(errno));
-      return 1;
-    }
-    close(listen_sock);
-    // Ensure that the socket is blocking
-    int fd_flags = fcntl(rvfi_dii_sock, F_GETFL);
-    if (fd_flags == -1) {
-      fprintf(stderr, "Failed to get file descriptor flags for socket!\n");
-      return 1;
-    }
-    if (config_print_rvfi) {
-      fprintf(stderr, "RVFI socket fd flags=%d, nonblocking=%d\n", fd_flags,
-              (fd_flags & O_NONBLOCK) != 0);
-    }
-    if (fd_flags & O_NONBLOCK) {
-      fprintf(stderr, "Socket was non-blocking, this will not work!\n");
-      return 1;
-    }
-    printf("Connected\n");
-  } else
-    entry = load_sail(initial_elf_file, /*main_file=*/true);
-#else
-  uint64_t entry = load_sail(initial_elf_file, /*main_file=*/true);
-#endif
+  if (rvfi) {
+    int exit_code = rvfi->setup_socket(config_print_rvfi);
+    if (exit_code)
+      exit(exit_code);
```

Comment:
```suggestion
    if (int exit_code = rvfi->setup_socket(config_print_rvfi))
      return exit_code;
```
We are in main, so can just return?